subsys/bluetooth/mesh/light_ctl_cli.c

```diff
@@ -0,0 +1,380 @@
+
+	if (buf->len == 5) {
+		status.target.temp = net_buf_simple_pull_le16(buf);
+		status.target.delta_uv = net_buf_simple_pull_le16(buf);
+		status.remaining_time =
+			model_transition_decode(net_buf_simple_pull_u8(buf));
+	} else {
+		status.target.temp = status.current.temp;
+		status.target.delta_uv = status.current.delta_uv;
+		status.remaining_time = 0;
+	}
+
+	if (model_ack_match(&cli->ack_ctx, BT_MESH_LIGHT_CTL_TEMP_STATUS,
+			    ctx)) {
+		struct bt_mesh_light_ctl_temp_status *rsp =
+			(struct bt_mesh_light_ctl_temp_status *)
+				cli->ack_ctx.user_data;
+		*rsp = status;
+		model_ack_rx(&cli->ack_ctx);
+	}
+
+	if (cli->handlers->temp_status) {
+		cli->handlers->temp_status(cli, ctx, &status);
+	}
+}
+
+static void default_status_handle(struct bt_mesh_model *model,
+				  struct bt_mesh_msg_ctx *ctx,
+				  struct net_buf_simple *buf)
+{
+	if (buf->len != BT_MESH_LIGHT_CTL_MSG_LEN_DEFAULT_MSG) {
+		return;
+	}
+
+	struct bt_mesh_light_ctl_cli *cli = model->user_data;
+	struct bt_mesh_light_ctl_params status;
+
+	status.light = net_buf_simple_pull_le16(buf);
+	status.temp = net_buf_simple_pull_le16(buf);
+	status.delta_uv = net_buf_simple_pull_le16(buf);
+
+	if (model_ack_match(&cli->ack_ctx, BT_MESH_LIGHT_CTL_DEFAULT_STATUS,
+			    ctx)) {
+		struct bt_mesh_light_ctl_params *rsp =
+			(struct bt_mesh_light_ctl_params *)
+				cli->ack_ctx.user_data;
+		*rsp = status;
+		model_ack_rx(&cli->ack_ctx);
+	}
+
+	if (cli->handlers->default_status) {
+		cli->handlers->default_status(cli, ctx, &status);
+	}
+}
+
+const struct bt_mesh_model_op _bt_mesh_light_ctl_cli_op[] = {
+	{ BT_MESH_LIGHT_CTL_STATUS, BT_MESH_LIGHT_CTL_MSG_MINLEN_STATUS,
+	  ctl_status_handle },
+	{ BT_MESH_LIGHT_CTL_TEMP_RANGE_STATUS,
+	  BT_MESH_LIGHT_CTL_MSG_LEN_TEMP_RANGE_STATUS,
+	  temp_range_status_handle },
+	{ BT_MESH_LIGHT_CTL_TEMP_STATUS,
+	  BT_MESH_LIGHT_CTL_MSG_MINLEN_TEMP_STATUS, temp_status_handle },
+	{ BT_MESH_LIGHT_CTL_DEFAULT_STATUS,
+	  BT_MESH_LIGHT_CTL_MSG_LEN_DEFAULT_MSG, default_status_handle },
+	BT_MESH_MODEL_OP_END,
+};
+
+static int bt_mesh_light_ctl_cli_init(struct bt_mesh_model *model)
+{
+	struct bt_mesh_light_ctl_cli *cli = model->user_data;
+
+	cli->model = model;
+	net_buf_simple_init(cli->pub.msg, 0);
+	model_ack_init(&cli->ack_ctx);
+
+	return 0;
+}
+
+const struct bt_mesh_model_cb _bt_mesh_light_ctl_cli_cb = {
+	.init = bt_mesh_light_ctl_cli_init,
+};
+
+static int get_msg(struct bt_mesh_light_ctl_cli *cli,
+		   struct bt_mesh_msg_ctx *ctx, void *rsp, uint16_t opcode,
+		   uint16_t ret_opcode)
+{
+	BT_MESH_MODEL_BUF_DEFINE(msg, opcode, BT_MESH_LIGHT_CTL_MSG_LEN_GET);
+	bt_mesh_model_msg_init(&msg, opcode);
+
+	return model_ackd_send(cli->model, ctx, &msg, &cli->ack_ctx, ret_opcode,
+			       rsp);
+}
+
+int bt_mesh_light_ctl_get(struct bt_mesh_light_ctl_cli *cli,
+			  struct bt_mesh_msg_ctx *ctx,
+			  struct bt_mesh_light_ctl_status *rsp)
+{
+	if (rsp == NULL) {
+		return -EFAULT;
```

Comment:
Typically `-EINVAL`

Reply:
It looks like this check should be removed. At least if the user should be able to use these setters/getters with or without blocking. If you dont mind I will change this by removing the error check and add `return model_ackd_send(cli->model, ctx, &msg, rsp ? &cli->ack_ctx : NULL, BT_MESH_LIGHT_TEMP_RANGE_STATUS, rsp);` instead

Reply:
Re: I discovered that my solution is not a great idea, and at the same time I discovered a potential bug-case. If MIN_RANGE = MAX_RANGE we will get a division be zero error from the conversion function for temp_to_lvl, so either we must prevent MIN_RANGE = MAX_RANGE, or handle the division by zero case in temp_to_lvl in some manner.

Reply:
Sounds good